## 1. 49999 New York taxi trips
<p><img style="float: right;margin:5px 20px 5px 1px; max-width:300px" src="https://assets.datacamp.com/production/project_496/img/taxi.jpg"> To drive a yellow New York taxi, you have to hold a "medallion" from the city's <em>Taxi and Limousine Commission</em>. Recently, one of those changed hands for over one million dollars, which shows how lucrative the job can be.</p>
<p>But this is the age of business intelligence and analytics! Even taxi drivers can stand to benefit from some careful investigation of the data, guiding them to maximize their profits. In this project, we will analyze a random sample of 49999 New York journeys made in 2013. We will also use regression trees and random forests to build a model that can predict the locations and times when the biggest fares can be earned.</p>
<p>Let's start by taking a look at the data!</p>

In [2]:
# Loading the tidyverse
library(tidyverse)

# Reading in the taxi data
taxi <- read_csv('datasets/taxi.csv')

# Taking a look at the first few rows in taxi
head(taxi)

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors
Rows: 49999 Columns: 7
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (1): medallion
dbl  (5): pickup_longitude, pickup_latitude, trip_time_in_secs, fare_amount,...
dttm (1): pickup_datetime

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


medallion                        pickup_datetime     pickup_longitude
1 4D24F4D8EF35878595044A52B098DFD2 2013-01-13 10:23:00 -73.94646       
2 A49C37EB966E7B05E69523D1CB7BE303 2013-01-13 04:52:00 -73.99827       
3 1E4B72A8E623888F53A9693C364AC05A 2013-01-13 10:47:00 -73.95346       
4 F7E4E9439C46B8AD5B16AB9F1B3279D7 2013-01-13 11:14:00 -73.98137       
5 A9DC75D59E0EA27E1ED328E8BE8CD828 2013-01-13 11:24:00 -73.96800       
6 19BF1BB516C4E992EA3FBAEDA73D6262 2013-01-13 10:51:00 -73.98502       
  pickup_latitude trip_time_in_secs fare_amount tip_amount
1 40.77273        600                8.0        2.5       
2 40.74041        840               18.0        0.0       
3 40.77586         60                3.5        0.7       
4 40.72473        720               11.5        2.3       
5 40.76000        240                6.5        0.0       
6 40.76341        540                8.5        1.7

## 2. Cleaning the taxi data
<p>As you can see above, the <code>taxi</code> dataset contains the times and price of a large number of taxi trips. Importantly we also get to know the location, the longitude and latitude, where the trip was started.</p>
<p>Cleaning data is a large part of any data scientist's daily work. It may not seem glamorous, but it makes the difference between a successful model and a failure. The <code>taxi</code> dataset needs a bit of polishing before we're ready to use it.</p>

In [3]:
taxi <- taxi %>%
   rename(long = pickup_longitude, lat = pickup_latitude) %>% 
   filter(fare_amount > 0 | tip_amount > 0) %>%
   mutate(total = log(fare_amount + tip_amount))

taxi

medallion                        pickup_datetime     long      lat     
1     4D24F4D8EF35878595044A52B098DFD2 2013-01-13 10:23:00 -73.94646 40.77273
2     A49C37EB966E7B05E69523D1CB7BE303 2013-01-13 04:52:00 -73.99827 40.74041
3     1E4B72A8E623888F53A9693C364AC05A 2013-01-13 10:47:00 -73.95346 40.77586
4     F7E4E9439C46B8AD5B16AB9F1B3279D7 2013-01-13 11:14:00 -73.98137 40.72473
5     A9DC75D59E0EA27E1ED328E8BE8CD828 2013-01-13 11:24:00 -73.96800 40.76000
6     19BF1BB516C4E992EA3FBAEDA73D6262 2013-01-13 10:51:00 -73.98502 40.76341
7     5F2EFC03B544635C9B0E7A4AA4FF9AC3 2013-01-13 12:53:00 -73.97295 40.79527
8     8DEB70907D00AA1D7FF5E2683240549B 2013-01-13 07:59:00 -73.96577 40.76530
9     E15F7CCB808DD15E0496D830D3DEDECE 2013-01-13 08:09:00 -73.94768 40.77507
10    0B3D3D51C78E944F68DC04209E86D5F7 2013-01-13 12:53:00 -73.98457 40.72488
11    3C16CFAD2B12F3508F7211C37F8F8B8F 2013-01-13 13:07:00 -73.97068 40.78506
12    5888835B75CF97CBE738A58484B12A6D 2013-01-13 12:31:00 -74.00164 40.74414
13    58A836AD639867DB949723BA2A941734 2013-01-13 13:37:00 -73.98800 40.74960
14    C01368C76C8DEFA6C678CF0D54AE87F0 2013-01-13 13:31:00 -73.99309 40.74755
15    F86299BD8DF9B0C90A4E20AD2A44EAF7 2013-01-13 14:31:00 -73.96832 40.78688
16    00D07524C1482FF5A3CB0932BE29003E 2013-01-13 13:02:00 -73.95131 40.81007
17    2232F3F8B6124B1947AEF62509A97DAE 2013-01-13 16:16:00 -73.98534 40.72374
18    85045374BBBFA1376CE5D64DED56B010 2013-01-13 13:14:00 -73.91682 40.76967
19    70F5983D94E1BA0B217E9F8F447D4382 2013-01-13 16:30:00 -73.97288 40.78684
20    3AC3AB85F6E59CB391253FA638B854AA 2013-01-13 13:24:00 -73.98768 40.71979
21    60DC999A38D953EE86AB81C23C9480E8 2013-01-13 16:03:00 -73.98846 40.73724
22    084ACA045413975E5FB960ECBD9C6588 2013-01-13 15:58:00 -73.96040 40.76165
23    AC496252AF3119662DEDBFD24A70E83B 2013-01-13 16:08:00 -73.98025 40.75132
24    F2773B3D7BC9C2A18B942644F43DD33D 2013-01-13 14:45:00 -74.01029 40.71158
25    9E800FF6BCF49308A6BC0678ED27499D 2013-01-13 17:29:00 -73.97753 40.74235
26    ED9B6E969C35E16314E6863A95D83B5F 2013-01-13 17:48:00 -73.99191 40.73557
27    4FBA078630428EFA5EEEF2E67A293464 2013-01-13 15:14:00 -73.87447 40.77406
28    FDD65AC3468B2A12AF0D3577D64FBF76 2013-01-13 19:15:00 -73.99110 40.69207
29    135A8EDC6EBD6F4397B7A6D281C75AB0 2013-01-13 16:59:00 -73.97753 40.76359
30    131949A57D5717A3E7112404CCDFF27B 2013-01-13 17:36:00 -73.97504 40.75832
⋮     ⋮                                ⋮                   ⋮         ⋮       
49969 0CE65EC42F3ABEB0BCC08F274703DDE7 2013-12-03 11:58:00 -73.98287 40.74546
49970 126A5559920CCC1F4AA7A7DFB137D328 2013-12-05 22:12:00 -74.00825 40.73770
49971 221F94E1FB935807635C1045866796CC 2013-12-03 12:31:00 -73.95893 40.77771
49972 0C4726D4E2AF94BF8FE2D23EFDA20917 2013-12-05 22:03:00 -73.99159 40.74441
49973 A6DB36B570BD59E08FBE76086C5EE662 2013-12-03 13:36:00 -73.98994 40.74703
49974 60C3AD7179183044E91FF3B9B90C1CAA 2013-12-03 11:50:00 -73.99038 40.72965
49975 A910CF5A84FFE5F10B8CCBC06CC5F944 2013-12-05 22:42:00 -73.99249 40.74290
49976 037CB433AB5895649F7A9EF37F767EF1 2013-12-05 22:49:00 -73.96532 40.77145
49977 798452AE1E4F97CE2B491F9590861FEA 2013-12-03 14:26:00 -73.95112 40.78284
49978 204BAB16D3382C5A5711068B28E624C2 2013-12-03 14:12:00 -73.97968 40.76131
49979 F3BA458FFB70903630ABF3332CB983F1 2013-12-03 14:39:00 -73.94083 40.79271
49980 5221BCE45F9FD34B709EB0882885B7AB 2013-12-05 23:21:00 -73.94922 40.80278
49981 2B3EFA3719EA953A656D5041750787EB 2013-12-03 13:43:00 -73.78947 40.64633
49982 B65ECEC404246E0A0370540B5FE24AAB 2013-12-05 23:40:00 -73.98241 40.75505
49983 8220D3BEF8BCF3C5DB50C66F32D9AB61 2013-12-05 23:50:00 -73.98628 40.74044
49984 87CDD10EC56CB55A6F92D872A16AECCD 2013-12-03 14:32:00 -73.98254 40.76761
49985 758FE3823459A49A24051D260415E866 2013-12-05 22:42:00 -73.98413 40.72923
49986 554389035D554151A222468199443C39 2013-12-06 00:35:00 -73.94059 40.71187
49987 1AABA654E5ABF4F25333847479904504 2013-12-03 13:55:00 -73.99451 40.74100
49988 0C4CEF33F6F0D06E62988C

## 3. Zooming in on Manhattan
<p>While the dataset contains taxi trips from all over New York City, the bulk of the trips are to and from Manhattan, so let's focus only on trips initiated there.</p>

In [4]:
# Reducing the data to taxi trips starting in Manhattan
# Manhattan is bounded by the rectangle with 
# latitude from 40.70 to 40.83 and 
# longitude from -74.025 to -73.93
taxi <- taxi  %>% 
    filter(between(lat,  40.70, 40.83) & 
           between(long, -74.025, -73.93))

taxi

medallion                        pickup_datetime     long      lat     
1     4D24F4D8EF35878595044A52B098DFD2 2013-01-13 10:23:00 -73.94646 40.77273
2     A49C37EB966E7B05E69523D1CB7BE303 2013-01-13 04:52:00 -73.99827 40.74041
3     1E4B72A8E623888F53A9693C364AC05A 2013-01-13 10:47:00 -73.95346 40.77586
4     F7E4E9439C46B8AD5B16AB9F1B3279D7 2013-01-13 11:14:00 -73.98137 40.72473
5     A9DC75D59E0EA27E1ED328E8BE8CD828 2013-01-13 11:24:00 -73.96800 40.76000
6     19BF1BB516C4E992EA3FBAEDA73D6262 2013-01-13 10:51:00 -73.98502 40.76341
7     5F2EFC03B544635C9B0E7A4AA4FF9AC3 2013-01-13 12:53:00 -73.97295 40.79527
8     8DEB70907D00AA1D7FF5E2683240549B 2013-01-13 07:59:00 -73.96577 40.76530
9     E15F7CCB808DD15E0496D830D3DEDECE 2013-01-13 08:09:00 -73.94768 40.77507
10    0B3D3D51C78E944F68DC04209E86D5F7 2013-01-13 12:53:00 -73.98457 40.72488
11    3C16CFAD2B12F3508F7211C37F8F8B8F 2013-01-13 13:07:00 -73.97068 40.78506
12    5888835B75CF97CBE738A58484B12A6D 2013-01-13 12:31:00 -74.00164 40.74414
13    58A836AD639867DB949723BA2A941734 2013-01-13 13:37:00 -73.98800 40.74960
14    C01368C76C8DEFA6C678CF0D54AE87F0 2013-01-13 13:31:00 -73.99309 40.74755
15    F86299BD8DF9B0C90A4E20AD2A44EAF7 2013-01-13 14:31:00 -73.96832 40.78688
16    00D07524C1482FF5A3CB0932BE29003E 2013-01-13 13:02:00 -73.95131 40.81007
17    2232F3F8B6124B1947AEF62509A97DAE 2013-01-13 16:16:00 -73.98534 40.72374
18    70F5983D94E1BA0B217E9F8F447D4382 2013-01-13 16:30:00 -73.97288 40.78684
19    3AC3AB85F6E59CB391253FA638B854AA 2013-01-13 13:24:00 -73.98768 40.71979
20    60DC999A38D953EE86AB81C23C9480E8 2013-01-13 16:03:00 -73.98846 40.73724
21    084ACA045413975E5FB960ECBD9C6588 2013-01-13 15:58:00 -73.96040 40.76165
22    AC496252AF3119662DEDBFD24A70E83B 2013-01-13 16:08:00 -73.98025 40.75132
23    F2773B3D7BC9C2A18B942644F43DD33D 2013-01-13 14:45:00 -74.01029 40.71158
24    9E800FF6BCF49308A6BC0678ED27499D 2013-01-13 17:29:00 -73.97753 40.74235
25    ED9B6E969C35E16314E6863A95D83B5F 2013-01-13 17:48:00 -73.99191 40.73557
26    135A8EDC6EBD6F4397B7A6D281C75AB0 2013-01-13 16:59:00 -73.97753 40.76359
27    131949A57D5717A3E7112404CCDFF27B 2013-01-13 17:36:00 -73.97504 40.75832
28    09EE700858482DA94CCBC192D2208B10 2013-01-13 15:36:00 -73.98092 40.75594
29    CDC40A1F194AA8D4CA9EF331A266E1FA 2013-01-13 17:39:00 -73.96726 40.76012
30    BA2E28EC4C10FA42A04131313286A591 2013-01-13 18:58:00 -73.97436 40.75417
⋮     ⋮                                ⋮                   ⋮         ⋮       
45737 C8980A6501C7D728020B1FC2FC1B64B1 2013-12-03 12:41:00 -73.98369 40.75583
45738 0CE65EC42F3ABEB0BCC08F274703DDE7 2013-12-03 11:58:00 -73.98287 40.74546
45739 126A5559920CCC1F4AA7A7DFB137D328 2013-12-05 22:12:00 -74.00825 40.73770
45740 221F94E1FB935807635C1045866796CC 2013-12-03 12:31:00 -73.95893 40.77771
45741 0C4726D4E2AF94BF8FE2D23EFDA20917 2013-12-05 22:03:00 -73.99159 40.74441
45742 A6DB36B570BD59E08FBE76086C5EE662 2013-12-03 13:36:00 -73.98994 40.74703
45743 60C3AD7179183044E91FF3B9B90C1CAA 2013-12-03 11:50:00 -73.99038 40.72965
45744 A910CF5A84FFE5F10B8CCBC06CC5F944 2013-12-05 22:42:00 -73.99249 40.74290
45745 037CB433AB5895649F7A9EF37F767EF1 2013-12-05 22:49:00 -73.96532 40.77145
45746 798452AE1E4F97CE2B491F9590861FEA 2013-12-03 14:26:00 -73.95112 40.78284
45747 204BAB16D3382C5A5711068B28E624C2 2013-12-03 14:12:00 -73.97968 40.76131
45748 F3BA458FFB70903630ABF3332CB983F1 2013-12-03 14:39:00 -73.94083 40.79271
45749 5221BCE45F9FD34B709EB0882885B7AB 2013-12-05 23:21:00 -73.94922 40.80278
45750 B65ECEC404246E0A0370540B5FE24AAB 2013-12-05 23:40:00 -73.98241 40.75505
45751 8220D3BEF8BCF3C5DB50C66F32D9AB61 2013-12-05 23:50:00 -73.98628 40.74044
45752 87CDD10EC56CB55A6F92D872A16AECCD 2013-12-03 14:32:00 -73.98254 40.76761
45753 758FE3823459A49A24051D260415E866 2013-12-05 22:42:00 -73.98413 40.72923
45754 554389035D554151A222468199443C39 2013-12-06 00:35:00 -73.94059 40.71187
45755 1AABA654E5ABF4F25333847479904504 2013-12-03 13:55:00 -73.99451 40.74100
45756 0C4CEF33F6F0D06E62988C

## 4. Where does the journey begin?
<p>It's time to draw a map! We're going to use the excellent <code>ggmap</code> package together with <code>ggplot2</code> to visualize where in Manhattan people tend to start their taxi journeys.</p>

In [ ]:
# Loading in ggmap and viridis for nice colors
library(ggmap)
library(viridis)

register_google(key = "AIzaSyCYm9kyz4Cjbt9JjNxcbUXZHNQvItQkj9c")

# Retrieving a stored map object which originally was created by
# nycmap <- get_map("manhattan", zoom = 12, color = "bw")
manhattan <- readRDS("datasets/manhattan.rds")

# Drawing a density map with the number of journey start locations
ggmap(manhattan, darken=0.5)+
    scale_fill_viridis(option='plasma') +
    geom_bin2d(data = taxi, aes(x = long, y = lat), bins = 60, alpha = 0.6) +
    labs(x = 'Longitude', y = 'Latitude', fill = 'Journeys')

ERROR: Error in readRDS("datasets/manhattan.rds"): formato de entrada desconocido


## 5. Predicting taxi fares using a tree
<p>The map from the previous task showed that the journeys are highly concentrated in the business and tourist areas. We also see that some taxi trips originating in Brooklyn slipped through, but that's fine. </p>
<p>We're now going to use a regression tree to predict the <code>total</code> fare with <code>lat</code> and <code>long</code> being the predictors. The <code>tree</code> algorithm will try to find cutpoints in those predictors that results in the decision tree with the best predictive capability.  </p>

In [ ]:
# Loading in the tree package
library(tree)

# Fitting a tree to lat and long
fitted_tree <- tree(total ~ lat + long, data = taxi)

# draw a diagram of the tree structure
plot(fitted_tree)
text(fitted_tree)

## 6. It's time. More predictors.
<p>The tree above looks a bit frugal, it only includes one split: It predicts that trips where <code>lat &lt; 40.7237</code> are more expensive, which makes sense as it is downtown Manhattan. But that's it. It didn't even include <code>long</code> as <code>tree</code> deemed that it didn't improve the predictions. Taxi drivers will need more information than this and any driver paying for your data-driven insights would be disappointed with that. As we know from Robert de Niro, it's best not to upset New York taxi drivers.</p>
<p>Let's start by adding some more predictors related to the <em>time</em> the taxi trip was made.</p>

In [ ]:
# Loading in the lubridate package
library(lubridate)

# Generate the three new time variables
taxi <- taxi %>% 
    mutate(hour = hour(pickup_datetime),
           wday = wday(pickup_datetime, label = TRUE),
           month = month(pickup_datetime, label = TRUE))

taxi

## 7. One more tree!
<p>Let's try fitting a new regression tree where we include the new time variables.</p>

In [ ]:
# Fitting a tree to lat and long
fitted_tree <- tree(total ~ lat + long + hour + wday + month, data = taxi)

# draw a diagram of the tree structure
plot(fitted_tree)
text(fitted_tree)

# Summarizing the performance of the tree
summary(fitted_tree)

## 8. One tree is not enough
<p>The regression tree has not changed after including the three time variables. This is likely because latitude is still the most promising first variable to split the data on, and after that split, the other variables are not informative enough to be included. A random forest model, where many different trees are fitted to subsets of the data, may well include the other variables in some of the trees that make it up. </p>

In [ ]:
# Loading in the randomForest package
library(randomForest)

# Fitting a random forest
fitted_forest <- randomForest(total ~ lat + long + hour + wday + month,
    data=taxi, ntree=80, sampsize=10000)

# Printing the fitted_forest object
fitted_forest

## 9. Plotting the predicted fare
<p>In the output of <code>fitted_forest</code> you should see the <code>Mean of squared residuals</code>, that is, the average of the squared errors the model makes. If you scroll up and check the <code>summary</code> of <code>fitted_tree</code> you'll find <code>Residual mean deviance</code> which is the same number. If you compare these numbers, you'll see that <code>fitted_forest</code> has a slightly lower error. Neither predictive model is <em>that</em> good, in statistical terms, they explain only about 3% of the variance. </p>
<p>Now, let's take a look at the predictions of <code>fitted_forest</code> projected back onto Manhattan.</p>

In [ ]:
# Extracting the prediction from forest_fit
taxi$pred_total <- fitted_forest$predicted

# Plotting the predicted mean trip prices from according to the random forest
ggmap(manhattan, darken=0.5) +
    scale_fill_viridis(option = 'plasma') +
    stat_summary_2d(data=taxi, aes(x = long, y = lat, z = pred_total),
                    fun = mean, alpha = 0.6, bins = 60) +
    labs(x = 'Longitude', y = 'Latitude', fill = 'Log fare+tip')

## 10. Plotting the actual fare
<p>Looking at the map with the predicted fares we see that fares in downtown Manhattan are predicted to be high, while midtown is lower. This map only shows the prediction as a function of <code>lat</code> and <code>long</code>, but we could also plot the predictions over time, or a combination of time and space, but we'll leave that for another time.</p>
<p>For now, let's compare the map with the predicted fares with a new map showing the mean fares according to the data.</p>

In [ ]:
# Function that returns the mean *if* there are 15 or more datapoints
mean_if_enough_data <- function(x) { 
    ifelse( length(x) >= 15, mean(x), NA) 
}

# Plotting the mean trip prices from the data
ggmap(manhattan, darken=0.5) +
    stat_summary_2d(data=taxi, aes(x = long, y = lat, z = total),
                    fun = mean_if_enough_data,
                    alpha = 0.6, bins = 60) +
  scale_fill_viridis(option = 'plasma') +
  labs(x = 'Longitude', y = 'Latitude', fill = 'Log fare+tip')

## 11. Where  do people spend the most?
<p>So it looks like the random forest model captured some of the patterns in our data. At this point in the analysis, there are many more things we could do that we haven't done. We could add more predictors if we have the data. We could try to fine-tune the parameters of <code>randomForest</code>. And we should definitely test the model on a hold-out test dataset. But for now, let's be happy with what we have achieved!</p>
<p>So, if you are a taxi driver in NYC, where in Manhattan would you expect people to spend the most on a taxi ride?</p>

In [ ]:
# Where are people spending the most on their taxi trips?
spends_most_on_trips <- "downtown" # "uptown" or "downtown"